### ISSUE TITLE
Failure to cross-compile with mingw-w64 on macOS

### ISSUE THREAD

**Issue by zoain-it** [twistedfall/opencv-rust#133]
<img width="732" alt="截屏2020-05-25 下午1 53 30" src="https://user-images.githubusercontent.com/43292634/82782530-1f3fac80-9e8f-11ea-9ad8-5547e67d21fd.png">

<img width="732" alt="截屏2020-05-25 下午1 55 31" src="https://user-images.githubusercontent.com/43292634/82782650-69289280-9e8f-11ea-893f-048a457a90c8.png">

<img width="732" alt="截屏2020-05-25 下午1 56 08" src="https://user-images.githubusercontent.com/43292634/82782688-7e9dbc80-9e8f-11ea-9cb0-d9ef1a34a186.png">

<img width="732" alt="截屏2020-05-25 下午1 56 19" src="https://user-images.githubusercontent.com/43292634/82782697-84939d80-9e8f-11ea-9eb4-6acaf0f6804e.png">

<img width="732" alt="截屏2020-05-25 下午1 52 00" src="https://user-images.githubusercontent.com/43292634/82782459-f61f1c00-9e8e-11ea-8813-9a1816be1d51.png">

opencv is also cross-compiled on macOS using mingw-w64

**Comment by twistedfall** [comment]
Are you building with `buildtime-bindgen` feature enabled?

**Comment by zoain-it** [comment]
None

**Comment by twistedfall** [comment]
Please do :) For macOS it's required

**Comment by zoain-it** [comment]
<img width="1440" alt="截屏2020-05-25 下午7 28 10" src="https://user-images.githubusercontent.com/43292634/82809116-f08cfa80-9ebd-11ea-8db7-61bc15e0fbc3.png">

I added it and it looks like the error message is much less than before. But there are still 3 functions not found.

**Comment by twistedfall** [comment]
I see, those errors are familiar, I've added some workaround for them, but only for `target_env = "msvc"`, seems like it's more common. Can you please show me the output of:
```sh
rustc --print=cfg
```
on your machine?

**Comment by zoain-it** [comment]
<img width="732" alt="截屏2020-05-25 下午7 58 52" src="https://user-images.githubusercontent.com/43292634/82810901-2a600000-9ec2-11ea-9cb5-766c07afeb38.png">

Please see.

**Comment by zoain-it** [comment]
Do I need to set target_env manually? How do I set it up?

**Comment by twistedfall** [comment]
No, the rustc sets it internally, there is no way you can influence it. Did you install rust through rustup? Can you show me:
```sh
rustup show
```

**Comment by zoain-it** [comment]
<img width="482" alt="截屏2020-05-25 下午8 09 19" src="https://user-images.githubusercontent.com/43292634/82811561-a1e25f00-9ec3-11ea-9c6a-233e2ce8cbef.png">

**Comment by twistedfall** [comment]
Ah, I forgot about the cross-compile part. What is the target triple you're using for the build? And can you please show:
```sh
rustc --target <cross compile triple> --print=cfg
```

**Comment by zoain-it** [comment]
<img width="730" alt="截屏2020-05-25 下午8 15 32" src="https://user-images.githubusercontent.com/43292634/82811931-7dd34d80-9ec4-11ea-96df-a4af590f6ffd.png">

here. is gnu

**Comment by twistedfall** [comment]
I've enabled the workaround for any windows build. Can you check the branch named `windows-workaround` and tell if it works for you?

**Comment by zoain-it** [comment]
All right, hold on.

**Comment by zoain-it** [comment]
It still doesn't seem to work.

I even tried, still not working.
```rust
// if cfg!(target_os = "windows") {
    ...
// }
```
Here's how I quote.
<img width="1440" alt="截屏2020-05-25 下午9 53 42" src="https://user-images.githubusercontent.com/43292634/82819040-4e2b4200-9ed2-11ea-8ab3-c23e9d23f21e.png">

**Comment by twistedfall** [comment]
That's quite weird, and the error is the same like the last time or it fails linking with different symbols?

**Comment by zoain-it** [comment]
The same symbols.

**Comment by zoain-it** [comment]
Very sorry that your modifications worked. It should have just been my quote that failed, I'm not quite sure exactly which operation was in question. I just modified the official library directly:
```rust
// if cfg!(target_env = "msvc") {
    ...
// }
```
It's working fine.

**Comment by twistedfall** [comment]
I'm sorry, I'm not 100% sure what you're saying here. So does the branch `windows-workaround` work for you without any additional modifications?

**Comment by zoain-it** [comment]
The windows-workaround branch doesn't seem to work. I followed your changes, modified my local master branch library, and it works fine!

**Comment by twistedfall** [comment]
But your change that works was to comment out the `if cfg!(target_env = "msvc") {` or to replace it with `if cfg!(target_os = "windows") {`?

**Comment by zoain-it** [comment]
Sorry to keep you waiting, I just tested a lot of situations. What finally allowed me to work properly was that I deleted this judgment.

**Comment by zoain-it** [comment]
I can assure you I'm using the same tool chain.

**Comment by zoain-it** [comment]
It's weird. I am change `if cfg!(target_os = "windows") {` to `if cfg!(target_os = "macos") {`. He can work.

**Comment by zoain-it** [comment]
I used `cargo run --target "x86_64-pc-windows-gnu" -vv` to compile
<img width="456" alt="截屏2020-05-25 下午11 14 14" src="https://user-images.githubusercontent.com/43292634/82825242-77050480-9edd-11ea-8344-fa2e31370f3a.png">

**Comment by twistedfall** [comment]
Yeah, I kind of suspected that. I've added some debug info and pushed a new branch `windows-workaround2`, can you please try with it and show me the build log output starting with "=== Enabled features:"?

**Comment by zoain-it** [comment]
All right, hold on.

**Comment by zoain-it** [comment]
<img width="742" alt="截屏2020-05-25 下午11 24 47" src="https://user-images.githubusercontent.com/43292634/82826071-f9da8f00-9ede-11ea-8d07-ee63b9ff3ab4.png">
Everything looks normal.

**Comment by twistedfall** [comment]
I think I know what’s the problem. But it’s not trivial to fix. For now please use the workaround with commenting out those lines. But please keep the bug report open, I’ll ping you when there’s a solution.

**Comment by zoain-it** [comment]
I seem to have run into another problem. Compile and get windows running error. I can't seem to find anything about it on the search engines.
<img width="820" alt="截屏2020-05-26 上午11 05 26" src="https://user-images.githubusercontent.com/43292634/82856557-d6483080-9f40-11ea-938f-28824258be73.png">

**Comment by zoain-it** [comment]
I looked at the compilation information and it relies on libclang.dylib for cross compilation. I suspect it is the issue here.

**Comment by twistedfall** [comment]
Well, that particular error is coming from inside the OpenCV library (not the rust crate), I don't really have any control over that. From the looks of it, it might be a problem between processor capabilities between the host machine (AVX2 present) and the crosscompilation target machine (AVX2 absent) and when you did a crosscompilation of OpenCV library on your macOS host machine it picked up the processor capabilities and baked them into the resulting library file. You might try to rebuild the crosscompiled OpenCV on the host machine while manually disabling some processor features. Please check here: https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

**Comment by zoain-it** [comment]
Okay, thanks. I worked it out.

**Comment by twistedfall** [comment]
Can you please check the `windows-fix` branch to see if it fixes the initial problem with crosscompiling for Windows?

**Comment by zoain-it** [comment]
All right, hold on.

**Comment by zoain-it** [comment]
It's great. He can work.

**Comment by zoain-it** [comment]
But there is another problem with the mingw tool chain. `core.o: too many sections` and `core.o: file too big`.
<img width="729" alt="截屏2020-05-26 下午10 38 36" src="https://user-images.githubusercontent.com/43292634/82915515-6cad3e00-9fa3-11ea-95d3-570194deb015.png">

This is a common problem in the MINGW toolchain. I modified build.rs and added `-Wa,-mbig-obj` to solve this problem.

**Comment by twistedfall** [comment]
I wasn't aware of that, I'll try to incorporate it into the crate too, thanks!

**Comment by zoain-it** [comment]
You're welcome. I should thank you for helping me out.

**Comment by twistedfall** [comment]
I've pushed the commit with the `big-obj` option to the `windows-fix` branch, can you check out if it fixes the problem for you?

**Comment by zoain-it** [comment]
He can work, great.

**Comment by twistedfall** [comment]
Fixed in 0.38
